tools/pinning/DESIGN.md

```diff
@@ -0,0 +1,124 @@
+specifying the conditions where a dependency is required instead of dynamically
+generating the list of required dependencies at runtime. This static string
+keeps the package's declaration of its dependencies consistent across
+environments.
+
+The other relatively recent change in Python packaging is the adoption of
+[pyproject.toml files](https://www.python.org/dev/peps/pep-0518/) which allows
+sdists to define their packages using a static file instead of a setup.py
+file which I expect most Python developers are more familiar with as I'm
+writing this. Using a static file instead of arbitrary Python code makes it
+much easier for package declarations to be reliably interpreted. The
+introduction of pyproject.toml also allows for the use of build systems other
+than setuptools which becomes relevant in the next section of this doc.
+
+## Our pinning system
+
+### Overview
+
+The files under this directory are used to generate Certbot's pinning files.
+The files under `oldest` are used to generate the constraints file used for our
+"oldest" tests while `normal` is used to generate the constraints used
+everywhere else. `common` includes shared files that are used for both sets of
+pinnings.
+
+Under `normal` and `oldest`, there are two files as of writing this. One is a
+pyproject.toml file for use with [Poetry](https://python-poetry.org/) while
+the other is a script that can be run to regenerate pinnings. The
+pyproject.toml file defines a dummy Python package that depends on everything
+we want to pin. This file largely just depends on our own local packages,
+however, extra dependencies can be declared to further constrain package
+versions or to declare additional dependencies.
+
+The reason we use Poetry is that it is somewhat unique among Python packaging
+tools in that when locking dependencies, it makes a best effort approach to do
+this for all environments rather than just the current environment. This
+includes scanning the dependencies for all wheels and sdists of the package and
+recursively resolving dependencies declared through environment markers that
+are not relevant for the current platform. You can see this in action through
+the inclusion of dependencies like pywin32 which we only have a dependency on
+for Windows.
+
+### Potential problems
+
+As of writing this, I'm aware of two potential problems with this pinning
+system. The first is largely described earlier in the doc which is the problem
+of sdists that use code to dynamically declare its dependencies. It's simply
+not feasible to ensure this arbitrary Python code declares its dependencies in
+the same way across all environments. Luckily, this is a largely a theoretical
+problem and I'm aware of no issues with our current dependencies.
+
+The second problem with this approach is that build dependencies are not
+tracked and pinned. Unfortunately, [this seems to be a largely unsolved problem
+in Python right
```

Comment:
Once we know we have a build dependency on a package, we can just declare a runtime dependency in `pyproject.toml` to pin it like you suggest. We currently do this for packages like `cython` in our current `pyproject.toml` files.

I think the problem though is that there doesn't currently seem to be any tooling to automate the process of tracking down and listing what all the build dependencies of your project are. I think this is especially true if you want to list the build dependencies of your project for all environments as I think build dependencies have the same problems here that runtime dependencies do.

(I think this is further complicated by most Python build tools following the recommendation [here](https://www.python.org/dev/peps/pep-0517/#recommendations-for-build-frontends-non-normative) which says that build dependencies should be installed in an isolated environment with one of the benefits of that being you can have conflicting build dependencies among your packages. I think how to pin build dependencies if you want to allow conflicting build dependencies like this is kind of an open question.)

Reply:
Ah, yes, I see now, I didn't realize that it was doing some sort of *runtime* dependency resolution. I still suspect there's some hackish way to do it, but that would probably both involve even more "dummy" nonsense and breaking recommendations, so I _guess_ we don't need to include a solution for this unsolved python project as an add-on to this PR.

Reply:
So my understanding is it's certainly technically possible, however, I think it would involve modifying or building on the code of existing Python packaging tools. I've searched around about this problem a lot and I asked about it [here](https://discuss.python.org/t/pinning-build-dependencies/8363) where I think CPython and pip maintainers essentially told me that it's an unsolved problem.

I certainly could be missing or misunderstanding something, but I don't think I am. I nerded out on this project pretty hard, so if you'd like to read/hear more, I'm more than happy to share or if you want to dig into it yourself a bit, I'd love to learn more.

Regardless though, the pinning of build dependencies was also a problem with the current system in master and while I'm certainly biased, I personally think this PR improves things in a number of other ways.

Reply:
No; I trust you -- if that last comment wasn't clear, it was in jest. I marked this discussion as resolved, I think you've explained enough that I have a small sense of why this is hard, though I don't yet know all of the details.

Reply:
Thanks for clarifying. I didn't initially read it that way.